packages/plugin-ext/src/plugin/in-plugin-filesystem-watcher-proxy.ts

```diff
@@ -0,0 +1,150 @@
+    }
+
+    onFileSystemEvent(id: string, uri: URI, type: FileChangeEventType) {
+        const perSubscriberEventEmitter: Emitter<FileSystemEvent> | undefined = this.subscribers.get(id);
+        if (perSubscriberEventEmitter) {
+            perSubscriberEventEmitter.fire({ uri, type });
+        } else {
+            // shouldn't happen
+            // if it happans then a message was lost, unsubscribe to make state consistent
+            this.proxy.$unregisterFileSystemWatcher(id);
+        }
+    }
+}
+
+class FileSystemWatcher implements theia.FileSystemWatcher {
+    private subscriberData: SubscriberData;
+
+    private isIgnoreCreateEvents: boolean;
+    private isIgnoreChangeEvents: boolean;
+    private isIgnoreDeleteEvents: boolean;
+
+    private onDidCreateEmitter: Emitter<theia.Uri>;
+    private onDidChangeEmitter: Emitter<theia.Uri>;
+    private onDidDeleteEmitter: Emitter<theia.Uri>;
+
+    constructor(
+        subscriberData: SubscriberData,
+        ignoreCreateEvents: boolean = false,
+        ignoreChangeEvents: boolean = false,
+        ignoreDeleteEvents: boolean = false
+    ) {
+        this.isIgnoreCreateEvents = ignoreCreateEvents;
+        this.isIgnoreChangeEvents = ignoreChangeEvents;
+        this.isIgnoreDeleteEvents = ignoreDeleteEvents;
+
+        this.onDidCreateEmitter = new Emitter<theia.Uri>();
+        this.onDidChangeEmitter = new Emitter<theia.Uri>();
+        this.onDidDeleteEmitter = new Emitter<theia.Uri>();
+
+        this.subscriberData = subscriberData;
+        subscriberData.event((event: FileSystemEvent) => {
+            // Here ignore event flags are not analyzed because all the logic is
+            // moved to server side to avoid unneded data transfer via network.
+            // The flags are present just to be read only accesible for user.
+            switch (event.type) {
+                case 'updated':
+                    this.onDidChangeEmitter.fire(event.uri);
+                    break;
+                case 'created':
+                    this.onDidCreateEmitter.fire(event.uri);
+                    break;
+                case 'deleted':
+                    this.onDidDeleteEmitter.fire(event.uri);
+                    break;
+            }
+        });
+    }
+
+    get ignoreCreateEvents(): boolean {
+        return this.isIgnoreCreateEvents;
+    }
+
+    get ignoreChangeEvents(): boolean {
+        return this.isIgnoreChangeEvents;
+    }
+
+    get ignoreDeleteEvents(): boolean {
+        return this.isIgnoreDeleteEvents;
+    }
+
+    get onDidCreate(): Event<theia.Uri> {
+        return this.onDidCreateEmitter.event;
+    }
+
+    get onDidChange(): Event<theia.Uri> {
+        return this.onDidChangeEmitter.event;
+    }
+
+    get onDidDelete(): Event<theia.Uri> {
+        return this.onDidDeleteEmitter.event;
+    }
+
+    dispose(): void {
+        if (this.subscriberData.unsubscribe) {
```

Comment:
Up to you, but it would be nice dispose onDidCreateEmitter, onDidUpdateEmitter, onDidDeleteEmitter too. Example: https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/connection-status-service.ts#L100
Also you could create  dispose collection `protected readonly toDispose = new DisposableCollection();` and push all Emitters here, and then toDispose.dispose() utilize each of them.

Reply:
Thank you, will fix.